add separate singleFstats option to all ComputeFstat classes

dbkeitel opened the pull request:
In testing #346 I noticed that the handling of single-detector F-stats (`twoFX`) didn't really make sense. It was impossible to get them without setting `BSGL`, and the corresponding function didn't fail cleanly if one tried anyway.

 - add separate singleFstats option to all ComputeFstat classes
   (core classes and grid-based searches; no MCMC classes seem to currently make use of these attributes)
 - BSGL implies this
 - but can also use it on its own
 - always pre-initialise self.twoFX
   but only set FSTATQ_2F_PER_DET
   if self.singleFstats
 - get_fullycoherent_single_IFO_twoFs()
   now fails if not self.singleFstats

Rodrigo-Tenorio commented on pyfstat/core.py:
```diff
@@ -1012,9 +1021,10 @@ def get_fullycoherent_detstat(
         if not self.transientWindowType:
+            if self.singleFstats:
+                self.get_fullycoherent_single_IFO_twoFs()
             if self.BSGL is False:
```
```suggestion
            if not self.BSGL:
```

Rodrigo-Tenorio commented on pyfstat/core.py:
```diff
@@ -1768,10 +1783,11 @@ def get_semicoherent_det_stat(
 
+        if self.singleFstats:
+            self.get_semicoherent_single_IFO_twoFs(record_segments)
         if self.BSGL is False:
```
```suggestion
        if not self.BSGL:
```

Rodrigo-Tenorio commented on pyfstat/grid_based_searches.py:
```diff
@@ -131,7 +131,7 @@ def __init__(
         self.output_keys = self.search_keys.copy()
         self.output_keys.append("twoF")
-        if hasattr(self.search, "twoFX"):
+        if self.search.singleFstats:
```
Let me check if I got this right: `singleFstats` is the boolean label.

dbkeitel replied:
Yes. I named it for parallelism with CFSv2's `outputSingleFstats`, but dropping the verb since we're not necessarily outputting anything here...  
By contrast, the results are `self.twoFX`.

This whole mess could be reconsidered by making the attributes more hierarchical, e.g. having two dicts `self.results` and `self.options`. But I guess that falls under "PyFstat 2.0".

Rodrigo-Tenorio replied:
Yes

Rodrigo-Tenorio reviewed (changes requested):
See inline comments. Overall fine

dbkeitel commented:
Added test coverage by expanding the existing `TestComputeFstat::test_get_fully_coherent_BSGL` and on the way also
* added `randSeed` to `ComputeFstat` to make the `noiseSqrtSX` on-the-fly data generation reproducible
* store BSGL-related quantities to `self` (long-term these need to be come user variables anyway, see #260)

dbkeitel commented:
...now it might actually pass...

Rodrigo-Tenorio commented on pyfstat/core.py:
```diff
@@ -626,6 +632,10 @@ def init_computefstatistic(self):
             ] = self.injectSqrtSX
         else:
             FstatOAs.injectSqrtSX = lalpulsar.FstatOptionalArgsDefaults.injectSqrtSX
+        if hasattr(self, "randSeed") and self.randSeed is not None:
+            FstatOAs.randSeed = self.randSeed
+        else:
+            FstatOAs.randSeed = lalpulsar.FstatOptionalArgsDefaults.randSeed
```
```suggestion
        FstatOAs.randSeed = (
              getattr(self, "randSeed", None) or lalpulsar.FstatOptionalArgsDefaults.randSeed
       )
```


I agree with this. I tried to simplify it, but I *don't* want to prevent 0 from being used as a proper seed.
For the sake of it, I'll leave it as a suggestion.

Q: Is there any reason not to use 0 as a seed in a random number generator?

dbkeitel replied:
Actually that's a good point: I just checked and the corresponding CFSv2 help is `pecify random-number seed for reproducible noise (0 means use /dev/urandom for seeding).`, and indeed that's what the low-level code does: https://lscsoft.docs.ligo.org/lalsuite/lalpulsar/_generate_pulsar_signal_8c_source.html#l01002

So there actually wouldn't be a problem if we interpret a user-given `0` as "give me the default" at the same level as a `None`, and our documentation should be explicit about that too.

dbkeitel replied:
See latest commit.

Rodrigo-Tenorio replied:
Added a bit more verbosity to the docstring. Cool!

Rodrigo-Tenorio reviewed (dismissed):
Looks fine for now.

Rodrigo-Tenorio reviewed (approved):
Just included an explicit reference into the docstring pointing to XLALAddGaussianNoise().

dbkeitel commented:
Thanks. Whitespace fixed, merging once linter is happy.
